activerecord/lib/active_record/associations/collection_proxy.rb

```diff
@@ -1110,6 +1111,7 @@ def reload
       def reset
         proxy_association.reset
         proxy_association.reset_scope
+        @offsets = {}
         reset_scope
```

Comment:
You're right. My assumption here was wrong. Based on previous versions of Rails we would expect to get different objects even without calling reset/reload. I will update the PR.